atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/internalschema/TimestampToTransactionSchemaMap.java

```diff
@@ -0,0 +1,174 @@
+/*
+ * (c) Copyright 2018 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.internalschema;
+
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.immutables.value.Value;
+
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.fasterxml.jackson.databind.annotation.JsonSerialize;
+import com.google.common.collect.ImmutableRangeMap;
+import com.google.common.collect.Range;
+import com.google.common.collect.RangeMap;
+import com.google.common.collect.RangeSet;
+import com.google.common.collect.TreeRangeSet;
+import com.palantir.common.annotation.Output;
+import com.palantir.logsafe.SafeArg;
+import com.palantir.logsafe.exceptions.SafeIllegalArgumentException;
+
+/**
+ * A {@link TimestampToTransactionSchemaMap} keeps track of a mapping of timestamp ranges to transactions schema
+ * versions. The reason for using such a mapping is that clients may switch between mappings over time for various
+ * reasons.
+ *
+ * {@link TimestampToTransactionSchemaMap#timestampToTransactionsTableSchemaVersion()} is always expected to cover
+ * all timestamps. That is, the ranges present should span the range [1, +∞) and be connected.
```

Comment:
I've added some explanation:

```
 * Note that this does not mean that future behaviour is fixed, since in practice this map is often read as part of
 * a {@link com.palantir.atlasdb.coordination.ValueAndBound}. Thus, even though the map will contain as a key some
 * range [C, +∞) for a constant C, it should be valid up to some point V > C - and behaviour at timestamps after V
 * may subsequently be changed.
```

Reply:
Looks good - maybe remove the first often to make this firmer.